File: src/providers/openai-compatible.ts
/**
 * OpenAI 兼容接口适配器
 * 支持所有兼容 OpenAI API 格式的模型供应商
 */

import type {
  ModelProvider,
  Message,
  ToolDefinition,
  ModelConfig,
  ModelResponse
} from './types.js';

export class OpenAICompatibleProvider implements ModelProvider {
  name: string;
  private apiKey: string;
  private baseUrl: string;
  private defaultModel: string;

  constructor(
    name: string,
    apiKey: string,
    baseUrl: string,
    defaultModel: string
  ) {
    this.name = name;
    this.apiKey = apiKey;
    this.baseUrl = baseUrl.replace(/\/$/, ''); // 移除末尾斜杠
    this.defaultModel = defaultModel;
  }

  async chat(
    messages: Message[],
    tools?: ToolDefinition[],
    config?: ModelConfig
  ): Promise<ModelResponse> {
    const response = await fetch(`${this.baseUrl}/chat/completions`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${this.apiKey}`
      },
      body: JSON.stringify({
        model: config?.model || this.defaultModel,
        messages,
        tools: tools && tools.length > 0 ? tools : undefined,
        temperature: config?.temperature ?? 0.7,
        max_tokens: config?.max_tokens ?? 4096,
        top_p: config?.top_p
      })
    });

    if (!response.ok) {
      const error = await response.json().catch(() => ({})) as any;
      throw new Error(
        `${this.name} API error: ${error.error?.message || response.statusText}`
      );
    }

    const data = await response.json() as any;
    const choice = data.choices[0];

    return {
      content: choice.message.content,
      tool_calls: choice.message.tool_calls,
      finish_reason: choice.finish_reason === 'tool_calls' ? 'tool_calls' : 'stop',
      usage: data.usage
    };
  }

  async *chatStream(
    messages: Message[],
    tools?: ToolDefinition[],
    config?: ModelConfig
  ): AsyncIterable<string> {
    const response = await fetch(`${this.baseUrl}/chat/completions`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${this.apiKey}`
      },
      body: JSON.stringify({
        model: config?.model || this.defaultModel,
        messages,
        tools: tools && tools.length > 0 ? tools : undefined,
        temperature: config?.temperature ?? 0.7,
        max_tokens: config?.max_tokens ?? 4096,
        stream: true
      })
    });

    if (!response.ok) {
      throw new Error(`${this.name} API error: ${response.statusText}`);
    }

    const reader = response.body?.getReader();
    if (!reader) throw new Error('No response body');

    const decoder = new TextDecoder();
    let buffer = '';

    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      buffer += decoder.decode(value, { stream: true });
      const lines = buffer.split('\n');
      buffer = lines.pop() || '';

      for (const line of lines) {
        if (line.startsWith('data: ')) {
          const data = line.slice(6);
          if (data === '[DONE]') return;

          try {
            const parsed = JSON.parse(data);
            const content = parsed.choices[0]?.delta?.content;
            if (content) yield content;
          } catch {
            // 忽略解析错误
          }
        }
      }
    }
  }

  async listModels(): Promise<string[]> {
    const response = await fetch(`${this.baseUrl}/models`, {
      headers: {
        'Authorization': `Bearer ${this.apiKey}`
      }
    });

    if (!response.ok) {
      throw new Error(`Failed to list models: ${response.statusText}`);
    }

    const data = await response.json() as any;
    return data.data.map((model: { id: string }) => model.id);
  }
}

File: server/internal/providers/openai_compatible.go
package providers

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"

	"infringement-agent-server/internal/models"
)

// OpenAICompatibleProvider OpenAI 兼容接口适配器
type OpenAICompatibleProvider struct {
	name         string
	apiKey       string
	baseURL      string
	defaultModel string
	client       *http.Client
}

// NewOpenAICompatible 创建 OpenAI 兼容的 Provider
func NewOpenAICompatible(name, apiKey, baseURL, defaultModel string) *OpenAICompatibleProvider {
	return &OpenAICompatibleProvider{
		name:         name,
		apiKey:       apiKey,
		baseURL:      strings.TrimRight(baseURL, "/"),
		defaultModel: defaultModel,
		client:       &http.Client{},
	}
}

func (p *OpenAICompatibleProvider) Name() string {
	return p.name
}

// chatRequest OpenAI 请求体
type chatRequest struct {
	Model       string                  `json:"model"`
	Messages    []chatMessageForAPI     `json:"messages"`
	Tools       []models.ToolDefinition `json:"tools,omitempty"`
	Temperature float64                 `json:"temperature"`
	MaxTokens   int                     `json:"max_tokens"`
}

// chatMessageForAPI 发给 API 的消息格式
// Content 需要支持 string 和 array 两种序列化
type chatMessageForAPI struct {
	Role             models.MessageRole `json:"role"`
	Content          json.RawMessage    `json:"content"`
	ReasoningContent *string            `json:"reasoning_content,omitempty"` // Kimi K2.5 thinking 推理内容
	Name             string             `json:"name,omitempty"`
	ToolCalls        []models.ToolCall  `json:"tool_calls,omitempty"`
	ToolCallID       string             `json:"tool_call_id,omitempty"`
}

// convertMessagesToAPI 将内部 Message 转为 API 请求格式
func convertMessagesToAPI(messages []models.Message) []chatMessageForAPI {
	result := make([]chatMessageForAPI, 0, len(messages))
	for _, msg := range messages {
		apiMsg := chatMessageForAPI{
			Role:       msg.Role,
			Name:       msg.Name,
			ToolCalls:  msg.ToolCalls,
			ToolCallID: msg.ToolCallID,
		}

		// 传递 reasoning_content（Kimi K2.5 thinking 模式）
		if msg.ReasoningContent != "" {
			rc := msg.ReasoningContent
			apiMsg.ReasoningContent = &rc
		}

		// 序列化 Content（string 或 array）
		contentBytes, _ := json.Marshal(msg.Content)
		apiMsg.Content = contentBytes

		result = append(result, apiMsg)
	}
	return result
}

// chatResponseChoice OpenAI 响应选项
type chatResponseChoice struct {
	Message      choiceMessage `json:"message"`
	FinishReason string        `json:"finish_reason"`
}

type choiceMessage struct {
	Content          *string           `json:"content"`
	ReasoningContent *string           `json:"reasoning_content,omitempty"` // Kimi K2.5 thinking 推理内容
	ToolCalls        []models.ToolCall `json:"tool_calls,omitempty"`
}

type chatResponse struct {
	Choices []chatResponseChoice `json:"choices"`
	Usage   *models.Usage        `json:"usage,omitempty"`
	Error   *apiError            `json:"error,omitempty"`
}

type apiError struct {
	Message string `json:"message"`
}

func (p *OpenAICompatibleProvider) Chat(messages []models.Message, tools []models.ToolDefinition, config models.ModelConfig) (*models.ModelResponse, error) {
	model := config.Model
	if model == "" {
		model = p.defaultModel
	}

	reqBody := chatRequest{
		Model:       model,
		Messages:    convertMessagesToAPI(messages),
		Temperature: config.Temperature,
		MaxTokens:   config.MaxTokens,
	}
	if len(tools) > 0 {
		reqBody.Tools = tools
	}

	bodyBytes, err := json.Marshal(reqBody)
	if err != nil {
		return nil, fmt.Errorf("marshal request: %w", err)
	}

	req, err := http.NewRequest("POST", p.baseURL+"/chat/completions", bytes.NewReader(bodyBytes))
	if err != nil {
		return nil, fmt.Errorf("create request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+p.apiKey)

	resp, err := p.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("%s API request failed: %w", p.name, err)
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("read response: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		var errResp chatResponse
		_ = json.Unmarshal(respBody, &errResp)
		msg := resp.Status
		if errResp.Error != nil {
			msg = errResp.Error.Message
		}
		return nil, fmt.Errorf("%s API error: %s", p.name, msg)
	}

	var result chatResponse
	if err := json.Unmarshal(respBody, &result); err != nil {
		return nil, fmt.Errorf("unmarshal response: %w", err)
	}

	if len(result.Choices) == 0 {
		return nil, fmt.Errorf("%s: no choices in response", p.name)
	}

	choice := result.Choices[0]
	content := ""
	if choice.Message.Content != nil {
		content = *choice.Message.Content
	}
	reasoningContent := ""
	if choice.Message.ReasoningContent != nil {
		reasoningContent = *choice.Message.ReasoningContent
	}

	finishReason := "stop"
	if choice.FinishReason == "tool_calls" {
		finishReason = "tool_calls"
	}

	return &models.ModelResponse{
		Content:          content,
		ReasoningContent: reasoningContent,
		ToolCalls:        choice.Message.ToolCalls,
		FinishReason:     finishReason,
		Usage:            result.Usage,
	}, nil
}
